superchain-backend.md

```diff
@@ -0,0 +1,457 @@
+    - Standalone service.
+    - RPC server with method to check safety of single message (non-transitive safety).
+    - CLI config with RPCs.
+    - RPC client binding for fetching EVM logs of a block (using `eth_getLogs`) and block header data.
+    - Verbose executing-message parsing.
+- feat(interop): verifier safe head progression [`monorepo #9753`](https://github.com/ethereum-optimism/optimism/pull/9753) (Mar 6)
+    - Stacked on top of `#9612`.
+    - Adds superchain backend as sub-component of op-node.
+    - Adds superchain into derivation pipeline for safety-interface.
+    - Duplication of some superchain-backend message safety,
+      executing-message parsing and CLI code, possible due to git retarget issues.
+- wip(interop) block dependency graph [`monorepo #10044`](https://github.com/ethereum-optimism/optimism/pull/10044) (April 4)
+    - Stacked on top of `#9612`.
+    - `BlockDependencies` that holds on to safety information.
+    - Some doubts about the `dependents` map with key uniqueness / reorg safety.
+    - Global lock on data.
+    - Recursive block-safety algorithm, unable to handle loops.
+    - Centered around block-safety.
+    - Needs work for determining safety of initiating messages transitively efficiently.
+    - All in-memory, no back-pressure or limits.
+
+From the above we can use the draft of the standalone backend service, but will need to change:
+- The executing message parsing: detect from receipts instead
+- The safety-resolution: use an append-log model instead, to avoid recursion
+- Maintain only recent history in memory, flush the older history to disk, to not go OOM.
+- Revisit reorg protection: we do need checkpoints to determine how to roll back the log,
+  but should be able to do so without the current separate data mappings.
+
+# Proposed Solution
+
+<!-- *A high level overview of the proposed solution.
+When there are multiple alternatives there should be an explanation of why one solution was picked over other solutions.
+As a rule of thumb, including code snippets (except for defining an external API) is likely too low level.* -->
+
+## Closing draft PRs
+
+We close the specs PR and the 3 draft PRs.
+Parts of the implementation are still applicable, this can be cherry-picked with attribution.
+
+## Superchain Backend specs
+
+A new specs PR should describe the interface (without enshrining batching),
+as well as a simplified version of a standalone backend.
+
+Also note that the specs should be updated to cover the edge-cases of intra-block event ordering.
+An invariant for event-log index should be introduced.
+
+The interface:
+```
+type SafetyLevel # see design-doc
+
+type Verifier interface {
+    CheckMessage(Identifier, PayloadHash) SafetyLevel
+    CheckBlock(ChainID, BlockHash, BlockNumber) SafetyLevel
+}
+```
+
+Summary of simplified implementation to include in specs:
```

Comment:
How does this implementation avoid the temporary unresolved dependencies between blocks? If the L2 peer hasn't caught up yet for an initiating hash to be looked up. Does the resolution for the block requesting that initiating halt or rejects the resolution? If not does it somehow subscribe to needing that data at a later point?

Reply:
The append-logs can provide the data to resolve any initiating-messages. I'll update the summary to describe this, good catch.